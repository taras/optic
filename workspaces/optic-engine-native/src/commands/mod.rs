use chrono::{DateTime, TimeZone, Utc};
use std::process::Command;
use uuid::Uuid;

use cqrs_core::{Aggregate, AggregateCommand, AggregateEvent};
pub use serde::{Deserialize, Serialize};

pub mod endpoint;
pub mod rfc;
pub mod shape;

pub use endpoint::EndpointCommand;
pub use rfc::RfcCommand;
use shape::AddShape;
pub use shape::ShapeCommand;

use crate::events::rfc as rfc_events;
use crate::events::shape as shape_events;
use crate::events::{
  EndpointEvent, EventContext, RfcEvent, ShapeEvent, SpecEvent, WithEventContext,
};
use crate::projections::SpecProjection;
use crate::queries::history::HistoryQueries;
use crate::state::shape::ShapeKind;

#[derive(Deserialize, Debug, Clone, Serialize)]
#[serde(untagged)]
pub enum SpecCommand {
  EndpointCommand(EndpointCommand),
  RfcCommand(RfcCommand),
  ShapeCommand(ShapeCommand),
}

impl From<EndpointCommand> for SpecCommand {
  fn from(endpoint_command: EndpointCommand) -> Self {
    Self::EndpointCommand(endpoint_command)
  }
}

impl From<RfcCommand> for SpecCommand {
  fn from(rfc_command: RfcCommand) -> Self {
    Self::RfcCommand(rfc_command)
  }
}

impl From<ShapeCommand> for SpecCommand {
  fn from(shape_command: ShapeCommand) -> Self {
    Self::ShapeCommand(shape_command)
  }
}

// Errors
// ------

#[derive(Debug)]
pub enum SpecCommandError {
  Other(&'static str),
  Validation(String),
  Unimplemented(&'static str, SpecCommand),
}

impl std::fmt::Display for SpecCommandError {
  fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
    f.write_str(&self.to_string())
  }
}

impl std::error::Error for SpecCommandError {}

// CommandContext
// --------------

#[derive(Clone, Deserialize, Debug, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CommandContext {
  pub client_id: String,
  pub client_session_id: String,
  pub client_command_batch_id: String,
  pub created_at: DateTime<Utc>,
}

impl CommandContext {
  pub fn new(
    batch_id: String,
    client_id: String,
    client_session_id: String,
    created_at: DateTime<impl TimeZone>,
  ) -> Self {
    Self {
      client_id,
      client_command_batch_id: batch_id,
      client_session_id,
      created_at: created_at.with_timezone(&Utc),
    }
  }
}

impl Default for CommandContext {
  fn default() -> Self {
    Self {
      client_id: String::from("anonymous"),
      client_session_id: String::from("unknown-session"),
      client_command_batch_id: String::from("unknown-batch"),
      created_at: Utc.timestamp(0, 0),
    }
  }
}

// SpecCommandHandler
// ------------------

#[derive(Debug, Default)]
pub struct SpecCommandHandler {
  command_context: CommandContext,
  spec_projection: SpecProjection,
}

impl SpecCommandHandler {
  pub fn new(command_context: CommandContext, spec_projection: SpecProjection) -> Self {
    Self {
      command_context,
      spec_projection,
    }
  }

  pub fn with_command_context(&mut self, command_context: CommandContext) {
    self.command_context = command_context
  }
}

impl Aggregate for SpecCommandHandler {
  fn aggregate_type() -> &'static str {
    "spec-command-handler"
  }
}

impl AggregateEvent<SpecCommandHandler> for SpecEvent {
  fn apply_to(self, command_handler: &mut SpecCommandHandler) {
    command_handler.spec_projection.apply(self)
  }
}

impl<I> From<I> for SpecCommandHandler
where
  I: IntoIterator,
  I::Item: AggregateEvent<SpecProjection>,
{
  fn from(events: I) -> Self {
    let spec_projection = SpecProjection::from(events);
    Self::new(CommandContext::default(), spec_projection)
  }
}

// Command handling
// ----------------

impl AggregateCommand<SpecCommandHandler> for SpecCommand {
  type Error = SpecCommandError;
  type Event = SpecEvent;
  type Events = Vec<SpecEvent>;

  fn execute_on(self, command_handler: &SpecCommandHandler) -> Result<Self::Events, Self::Error> {
    let mut events = command_handler.spec_projection.execute(self)?;
    let event_context = EventContext::from(command_handler.command_context.clone());
    for event in &mut events {
      event.with_event_context(event_context.clone());
    }
    Ok(events)
  }
}

impl AggregateCommand<SpecProjection> for SpecCommand {
  type Error = SpecCommandError;
  type Event = SpecEvent;
  type Events = Vec<SpecEvent>;

  fn execute_on(self, spec_projection: &SpecProjection) -> Result<Self::Events, Self::Error> {
    let events = match self {
      SpecCommand::EndpointCommand(EndpointCommand::AddPathParameter(command)) => {
        let path_parameter_added_event = spec_projection
          .endpoint()
          .execute(EndpointCommand::AddPathParameter(command))?
          .into_iter()
          .find(|event| matches!(event, EndpointEvent::PathParameterAdded(_)))
          .expect("endpoint command handler should have failed if event wasnt produced");

        let path_parameter_added = match &path_parameter_added_event {
          EndpointEvent::PathParameterAdded(descriptor) => descriptor,
          _ => unreachable!(), // we already found the event of the pattern above or have panicked instead
        };

        let shape_added_event = ShapeEvent::from(ShapeCommand::add_shape(
          Uuid::new_v4().to_hyphenated().to_string(),
          ShapeKind::StringKind,
          String::from(""),
        ));

        let shape_id = match &shape_added_event {
          ShapeEvent::ShapeAdded(event) => event.shape_id.clone(),
          _ => unreachable!(),
        };

        let path_parameter_shape_set_event = EndpointEvent::from(
          EndpointCommand::set_path_parameter_shape(path_parameter_added.path_id.clone(), shape_id),
        );

        vec![
          SpecEvent::from(path_parameter_added_event),
          SpecEvent::from(shape_added_event),
          SpecEvent::from(path_parameter_shape_set_event),
        ]
      }

      SpecCommand::EndpointCommand(EndpointCommand::SetPathParameterShape(command)) => {
        spec_projection
          .shape()
          .execute(EndpointCommand::SetPathParameterShape(command.clone()))?; // validate shape exists
        let endpoint_events = spec_projection
          .endpoint()
          .execute(EndpointCommand::SetPathParameterShape(command.clone()))?;

        endpoint_events
          .into_iter()
          .map(|endpoint_event| SpecEvent::from(endpoint_event))
          .collect::<Vec<_>>()
      }

      SpecCommand::EndpointCommand(EndpointCommand::SetRequestBodyShape(command)) => {
        spec_projection
          .shape()
          .execute(EndpointCommand::SetRequestBodyShape(command.clone()))?;
        let endpoint_events = spec_projection
          .endpoint()
          .execute(EndpointCommand::SetRequestBodyShape(command))?;

        endpoint_events
          .into_iter()
          .map(|endpoint_event| SpecEvent::from(endpoint_event))
          .collect::<Vec<_>>()
      }

      SpecCommand::EndpointCommand(EndpointCommand::SetResponseBodyShape(command)) => {
        spec_projection
          .shape()
          .execute(EndpointCommand::SetResponseBodyShape(command.clone()))?;
        let endpoint_events = spec_projection
          .endpoint()
          .execute(EndpointCommand::SetResponseBodyShape(command))?;

        endpoint_events
          .into_iter()
          .map(|endpoint_event| SpecEvent::from(endpoint_event))
          .collect::<Vec<_>>()
      }

      // endpoint commands that can be purely handled by the endpoint projection
      SpecCommand::EndpointCommand(endpoint_command) => spec_projection
        .endpoint()
        .execute(endpoint_command)?
        .into_iter()
        .map(|endpoint_event| SpecEvent::from(endpoint_event))
        .collect::<Vec<_>>(),

      SpecCommand::RfcCommand(rfc_command) => spec_projection
        .history()
        .execute(rfc_command)?
        .into_iter()
        .map(|rfc_event| SpecEvent::from(rfc_event))
        .collect::<Vec<_>>(),

      SpecCommand::ShapeCommand(shape_command) => spec_projection
        .shape()
        .execute(shape_command)?
        .into_iter()
        .map(|shape_event| SpecEvent::from(shape_event))
        .collect::<Vec<_>>(),
    };
    Ok(events)
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use chrono::Local;
  use insta::assert_debug_snapshot;
  use serde_json::json;

  #[test]
  pub fn can_handle_add_path_parameter_command() {
    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
    ]))
    .expect("initial events should be valid spec events");

    let mut projection = SpecProjection::from(initial_events);

    let valid_command: SpecCommand = serde_json::from_value(json!(
      {"AddPathParameter": {"pathId": "path_2","parentPathId": "path_1","name": "todoId"}}
    ))
    .expect("example command should be a valid command");

    let new_events = projection
      .execute(valid_command)
      .expect("valid command should yield new events");
    assert_eq!(new_events.len(), 3);
    // TODO: figure out how insta redactions patterns work so we can assert a snapshot here
    // despite the generated uuid's (documentation is sparse :/)

    let unexisting_parent: SpecCommand = serde_json::from_value(json!(
      {"AddPathParameter": {"pathId": "path_2","parentPathId": "not-a-path","name": "todoId"}}
    ))
    .unwrap();
    let unexisting_parent_result = projection.execute(unexisting_parent);
    assert!(unexisting_parent_result.is_err());
    assert_debug_snapshot!(
      "can_handle_add_path_parameter_command__unexisting_parent_result",
      unexisting_parent_result.unwrap_err()
    );

    let unassignable_path_id: SpecCommand = serde_json::from_value(json!(
      {"AddPathParameter": {"pathId": "path_1","parentPathId": "path_1","name": "todoId"}}
    ))
    .unwrap();
    let unassignable_path_id_result = projection.execute(unassignable_path_id);
    assert!(unassignable_path_id_result.is_err());
    assert_debug_snapshot!(
      "can_handle_add_path_parameter_command__unassignable_path_id_result",
      unassignable_path_id_result.unwrap_err()
    );

    for event in new_events {
      projection.apply(event); // verify this doesn't panic goes a long way to verifying the events
    }
  }

  #[test]
  pub fn can_handle_set_path_parameter_shape_command() {
    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
      {"PathParameterAdded": {"pathId": "path_2","parentPathId": "path_1","name": "todoId"}},
      {"ShapeAdded":{"shapeId":"string_shape_1","baseShapeId":"$string","parameters":{"DynamicParameterList":{"shapeParameterIds":[]}},"name":"",}}
    ]))
    .expect("initial events should be valid spec events");

    let mut projection = SpecProjection::from(initial_events);

    let valid_command: SpecCommand = serde_json::from_value(json!(
      {"SetPathParameterShape":{"pathId":"path_2","shapedRequestParameterShapeDescriptor":{"shapeId":"string_shape_1","isRemoved":false}}}
    ))
    .expect("example command should be a valid command");

    let new_events = projection
      .execute(valid_command)
      .expect("valid command should yield new events");
    assert_eq!(new_events.len(), 1);

    let setting_root_path: SpecCommand = serde_json::from_value(json!(
      {"SetPathParameterShape":{"pathId":"root","shapedRequestParameterShapeDescriptor":{"shapeId":"string_shape_1","isRemoved":false}}}
    )).unwrap();
    let setting_root_path_result = projection.execute(setting_root_path);
    assert!(setting_root_path_result.is_err());
    assert_debug_snapshot!(
      "can_handle_set_path_parameter_shape_command__setting_root_path_result",
      setting_root_path_result.unwrap_err()
    );

    let unexisting_path: SpecCommand = serde_json::from_value(json!(
      {"SetPathParameterShape":{"pathId":"not-a-path","shapedRequestParameterShapeDescriptor":{"shapeId":"string_shape_1","isRemoved":false}}}
    )).unwrap();
    let unexisting_path_result = projection.execute(unexisting_path);
    assert!(unexisting_path_result.is_err());
    assert_debug_snapshot!(
      "can_handle_set_path_parameter_shape_command__unexisting_path_result",
      unexisting_path_result.unwrap_err()
    );

    let unexisting_shape: SpecCommand = serde_json::from_value(json!(
      {"SetPathParameterShape":{"pathId":"path_2","shapedRequestParameterShapeDescriptor":{"shapeId":"not_a_shape_id","isRemoved":false}}}
    )).unwrap();
    let unexisting_shape_result = projection.execute(unexisting_shape);
    assert!(unexisting_shape_result.is_err());
    assert_debug_snapshot!(
      "can_handle_set_path_parameter_shape_command__unexisting_shape_result",
      unexisting_shape_result.unwrap_err()
    );

    for event in new_events {
      projection.apply(event); // verify this doesn't panic goes a long way to verifying the events
    }
  }

  #[test]
  pub fn can_handle_set_request_body_shape_command() {
    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
      {"RequestAdded": {"requestId": "request_1", "pathId": "path_1", "httpMethod": "POST"}},
      {"ShapeAdded":{"shapeId":"string_shape_1","baseShapeId":"$string","parameters":{"DynamicParameterList":{"shapeParameterIds":[]}},"name":"",}}
    ]))
    .expect("initial events should be valid endpoint events");

    let mut projection = SpecProjection::from(initial_events);

    let valid_command: SpecCommand = serde_json::from_value(json!(
      {"SetRequestBodyShape": {"requestId": "request_1", "bodyDescriptor": { "httpContentType": "application/json", "shapeId": "string_shape_1", "isRemoved": false }}}
    ))
    .expect("example command should be a valid command");

    let new_events = projection
      .execute(valid_command)
      .expect("valid command should yield new events");
    assert_eq!(new_events.len(), 1);
    assert_debug_snapshot!(
      "can_handle_set_request_body_shape_command__new_events",
      new_events
    );

    let unexisting_shape: SpecCommand = serde_json::from_value(json!(
      {"SetRequestBodyShape": {"requestId": "request_1", "bodyDescriptor": { "httpContentType": "application/json", "shapeId": "not-a-shape-id", "isRemoved": false }}}
    ))
    .unwrap();
    let unexisting_shape_result = projection.execute(unexisting_shape);
    assert!(unexisting_shape_result.is_err());
    assert_debug_snapshot!(
      "can_handle_set_request_body_shape_command__unexisting_shape_result",
      unexisting_shape_result.unwrap_err()
    );

    for event in new_events {
      projection.apply(event); // verify this doesn't panic goes a long way to verifying the events
    }
  }

  #[test]
  pub fn can_handle_set_response_body_shape_command() {
    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
      {"RequestAdded": {"requestId": "request_1", "pathId": "path_1", "httpMethod": "POST"}},
      {"ResponseAddedByPathAndMethod": {"responseId": "response_1", "pathId": "path_1", "httpMethod": "GET", "httpStatusCode": 200}},
      {"ShapeAdded":{"shapeId":"string_shape_1","baseShapeId":"$string","parameters":{"DynamicParameterList":{"shapeParameterIds":[]}},"name":"",}}
    ]))
    .expect("initial events should be valid endpoint events");

    let mut projection = SpecProjection::from(initial_events);

    let valid_command: SpecCommand = serde_json::from_value(json!(
      {"SetResponseBodyShape": {"responseId": "response_1", "bodyDescriptor": { "httpContentType": "application/json", "shapeId": "string_shape_1", "isRemoved": false }}}
    ))
    .expect("example command should be a valid command");

    let new_events = projection
      .execute(valid_command)
      .expect("valid command should yield new events");
    assert_eq!(new_events.len(), 1);
    assert_debug_snapshot!(
      "can_handle_set_response_body_shape_command__new_events",
      new_events
    );

    let unexisting_shape: SpecCommand = serde_json::from_value(json!(
      {"SetResponseBodyShape": {"responseId": "response_1", "bodyDescriptor": { "httpContentType": "application/json", "shapeId": "not-a-shape-id", "isRemoved": false }}}
    ))
    .unwrap();
    let unexisting_shape_result = projection.execute(unexisting_shape);
    assert!(unexisting_shape_result.is_err());
    assert_debug_snapshot!(
      "can_handle_set_response_body_shape_command__unexisting_shape_result",
      unexisting_shape_result.unwrap_err()
    );

    for event in new_events {
      projection.apply(event); // verify this doesn't panic goes a long way to verifying the events
    }
  }

  #[test]
  pub fn spec_handler_provides_event_context_from_capture_context() {
    let initial_events: Vec<SpecEvent> = serde_json::from_value(json!([
      {"PathComponentAdded": {"pathId": "path_1","parentPathId": "root","name": "todos"}},
    ]))
    .expect("initial events should be valid spec events");

    let mut command_handler = SpecCommandHandler::from(initial_events);

    let command_context = CommandContext::new(
      String::from("test-batch-1"),
      String::from("test-client-id"),
      String::from("test-client-session-id"),
      Utc.timestamp(0, 0),
    );

    command_handler.with_command_context(command_context.clone());

    let valid_command: SpecCommand = serde_json::from_value(json!(
      {"AddPathComponent": {"pathId": "path_2","parentPathId": "path_1","name": "completed"}}
    ))
    .expect("example command should be a valid command");

    let new_events = command_handler
      .execute(valid_command)
      .expect("valid command should yield new events");

    let new_event = match new_events.get(0) {
      Some(SpecEvent::EndpointEvent(EndpointEvent::PathComponentAdded(event))) => event,
      _ => unreachable!(
        "PathComponentAdded event should have been generated from AddPathComponent command"
      ),
    };

    assert!(new_event.event_context.is_some());
    let event_context = new_event.event_context.clone().unwrap();

    assert_eq!(event_context.client_id, command_context.client_id);
    assert_eq!(
      event_context.client_session_id,
      command_context.client_session_id
    );
    assert_eq!(
      event_context.client_command_batch_id,
      command_context.client_command_batch_id
    );
    assert_eq!(
      event_context.created_at,
      command_context.created_at.to_rfc3339()
    );

    assert_debug_snapshot!(
      "spec_handler_provides_event_context_from_capture_context",
      &new_event
    );
  }
}
