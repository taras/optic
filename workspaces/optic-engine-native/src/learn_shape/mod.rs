use crate::shapes::JsonTrail;
use crate::BodyDescriptor;
use crate::HttpInteraction;
use std::collections::HashMap;

mod result;
mod traverser;
mod visitors;

pub use result::{TrailObservationsResult, TrailValues};
use traverser::Traverser;
use visitors::learn_json_values::LearnVisitors;
use visitors::BodyVisitors;

/// Traverse the description of a body and denote, per body trail, the types
/// of values seen.
pub fn observe_body_trails(
  into_body: impl Into<Option<BodyDescriptor>>,
) -> TrailObservationsResult {
  let body = into_body.into();
  let trail_map: HashMap<JsonTrail, TrailValues> = HashMap::new();

  let values_by_trail = if body.is_some() {
    let traverser = Traverser::new();
    let mut visitors = LearnVisitors::new();

    traverser.traverse_root_shape(body, &mut visitors);
    visitors.take_results()
  } else {
    HashMap::new()
  };

  TrailObservationsResult::from(values_by_trail)
}

#[cfg(test)]
mod test {
  use super::*;
  use insta::assert_debug_snapshot;
  use serde_json::json;

  #[test]
  fn trail_values_should_produce_magp() {
    let object_body = json!(
      [{"message": "hello"}, {"message": 123}, {"colors": ["red", true]}]
    );

    let body: Option<BodyDescriptor> = Some(BodyDescriptor::from(object_body));

    let result = observe_body_trails(body).normalized();

    let root_shape_trail = JsonTrail::empty();
    let root_shape_result = result
      .values()
      .find(|value| value.trail == root_shape_trail)
      .expect("results should contain a json trail for the root shape");
    assert!(root_shape_result.was_array);

    let item_trail = JsonTrail::empty().with_array_item(0);
    let item_shape_result = result
      .values()
      .find(|value| value.trail == item_trail)
      .expect("results should contain a json trail for the array item");
    dbg!(&item_shape_result);
    assert!(item_shape_result.was_object);
    assert_eq!(item_shape_result.field_sets.len(), 2);
  }
}
