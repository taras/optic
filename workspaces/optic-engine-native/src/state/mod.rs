use serde::de::{self, Deserializer, IgnoredAny, SeqAccess, Visitor};
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

pub mod body;
pub mod endpoint;
pub mod shape;

pub trait SpecIdGenerator {
  fn generate_id(&mut self, prefix: &str) -> String;

  fn field(&mut self) -> String {
    self.generate_id("field_")
  }

  fn request(&mut self) -> String {
    self.generate_id("request_")
  }

  fn response(&mut self) -> String {
    self.generate_id("response_")
  }

  fn shape(&mut self) -> String {
    self.generate_id("shape_")
  }

  fn shape_param(&mut self) -> String {
    self.generate_id("shape_param_")
  }
}

#[derive(Debug, Serialize)]
pub struct TaggedInput<T>(pub T, pub Tags);
pub type Tags = HashSet<String>;

impl<T> TaggedInput<T> {
  pub fn into_parts(self) -> (T, Tags) {
    (self.0, self.1)
  }

  pub fn parts(&self) -> (&T, &Tags) {
    (&self.0, &self.1)
  }

  pub fn into_input(self) -> T {
    self.0
  }
}

// Custom implementation of Deserialize, to allow ignoring additional items in the tuple,
// making it significantly easier to pipe results back in as inputs.
impl<'de, T> Deserialize<'de> for TaggedInput<T>
where
  T: Deserialize<'de>,
{
  fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
  where
    D: Deserializer<'de>,
  {
    struct TaggedInputVisitor<T> {
      marker: std::marker::PhantomData<T>,
    }

    impl<'de, T> Visitor<'de> for TaggedInputVisitor<T>
    where
      T: Deserialize<'de>,
    {
      type Value = TaggedInput<T>;

      fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
        formatter.write_str("struct TaggedInput")
      }

      fn visit_seq<V>(self, mut seq: V) -> Result<Self::Value, V::Error>
      where
        V: SeqAccess<'de>,
      {
        let s = seq
          .next_element()?
          .ok_or_else(|| de::Error::invalid_length(0, &self))?;
        let n = seq
          .next_element()?
          .ok_or_else(|| de::Error::invalid_length(1, &self))?;

        // we must visit the rest of the elements in the sequence to prevent panics
        while let Some(IgnoredAny) = seq.next_element()? {
          // but we can just ignore them
        }

        Ok(TaggedInput(s, n))
      }
    }

    deserializer.deserialize_seq(TaggedInputVisitor {
      marker: std::marker::PhantomData,
    })
  }
}

// pub struct TaggedInputIterator<T> {
//   inner: Option<TaggedInput<T>>,
// }

// impl<T> Iterator for TaggedInputIterator<T> {
//   type Item = (Tags, T);

//   #[inline]
//   fn next(&mut self) -> Option<Self::Item> {
//     self.inner.take();

//     tagged_input.map(|tagged_input| {
//       let TaggedInput(input, tags) = tagged_input;
//       (tags, input)
//     })
//   }
// }

// impl<T> IntoIterator for TaggedInput<T> {
//   type Item = (Tags, T);
//   type IntoIter = TaggedInputIterator<T>;

//   fn into_iter(self) -> Self::IntoIter {
//     TaggedInputIterator { inner: Some(self) }
//   }
// }
