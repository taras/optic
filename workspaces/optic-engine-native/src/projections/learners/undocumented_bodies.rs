use cqrs_core::{Aggregate, AggregateEvent, Event};
use serde::Serialize;
use std::collections::HashMap;

use crate::commands::{EndpointCommand, SpecCommand};
use crate::interactions::{BodyAnalysisLocation, BodyAnalysisResult};
use crate::learn_shape::TrailObservationsResult;
use crate::state::SpecIdGenerator;
use crate::JsonTrail;

#[derive(Default, Debug)]
pub struct LearnedUndocumentedBodiesProjection {
  observations_by_body_location: HashMap<BodyAnalysisLocation, TrailObservationsResult>,
}

impl LearnedUndocumentedBodiesProjection {
  fn with_body_analysis_result(&mut self, analysis: BodyAnalysisResult) {
    let existing_observations = self
      .observations_by_body_location
      .entry(analysis.body_location)
      .or_insert_with(|| TrailObservationsResult::default());

    existing_observations.union(analysis.trail_observations.normalized());
  }

  pub fn into_endpoint_bodies(
    self,
    id_generator: &mut impl SpecIdGenerator,
  ) -> impl Iterator<Item = EndpointBodies> {
    let mut endpoints_by_endpoint = HashMap::new();
    for (body_location, observations) in self.observations_by_body_location {
      let (root_shape_id, body_commands) =
        observations.into_commands(id_generator, &JsonTrail::empty());
      let endpoint_body =
        EndpointBody::new(&body_location, root_shape_id, body_commands, id_generator);

      let (path_id, method) = match body_location {
        BodyAnalysisLocation::UnmatchedRequest {
          path_id, method, ..
        } => (path_id, method),
        BodyAnalysisLocation::UnmatchedResponse {
          path_id, method, ..
        } => (path_id, method),
        _ => unreachable!("analyzing undocumented bodies should only yield unmatched results"),
      };

      let endpoint_bodies = endpoints_by_endpoint
        .entry((path_id, method))
        .or_insert_with_key(|(path_id, method)| {
          EndpointBodies::new(path_id.clone(), method.clone())
        });

      endpoint_bodies.push(endpoint_body);
    }

    endpoints_by_endpoint.into_iter().map(|(k, v)| v)
  }
}

impl Aggregate for LearnedUndocumentedBodiesProjection {
  fn aggregate_type() -> &'static str {
    "learned_undocument_bodies"
  }
}

impl Event for BodyAnalysisResult {
  fn event_type(&self) -> &'static str {
    "body_analysis_result"
  }
}

impl AggregateEvent<LearnedUndocumentedBodiesProjection> for BodyAnalysisResult {
  fn apply_to(self, aggregate: &mut LearnedUndocumentedBodiesProjection) {
    aggregate.with_body_analysis_result(self)
  }
}

// Output structs
// --------------

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointBodies {
  path_id: String,
  method: String,
  requests: Vec<EndpointRequestBody>,
  responses: Vec<EndpointResponseBody>,
}

impl EndpointBodies {
  pub fn new(path_id: String, method: String) -> Self {
    Self {
      path_id,
      method,
      requests: vec![],
      responses: vec![],
    }
  }

  pub fn push(&mut self, endpoint: EndpointBody) {
    match endpoint {
      EndpointBody::Request(endpoint_request) => {
        self.requests.push(endpoint_request);
      }
      EndpointBody::Response(endpoint_response) => {
        self.responses.push(endpoint_response);
      }
    }
  }
}

#[derive(Debug)]
pub enum EndpointBody {
  Request(EndpointRequestBody),
  Response(EndpointResponseBody),
}

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointRequestBody {
  commands: Vec<SpecCommand>,

  #[serde(skip)]
  path_id: String,
  #[serde(skip)]
  method: String,

  #[serde(flatten)]
  body_descriptor: Option<EndpointBodyDescriptor>,
}

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointResponseBody {
  commands: Vec<SpecCommand>,
  status_code: u16,

  #[serde(skip)]
  path_id: String,
  #[serde(skip)]
  method: String,

  #[serde(flatten)]
  body_descriptor: Option<EndpointBodyDescriptor>,
}

#[derive(Default, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EndpointBodyDescriptor {
  content_type: String,
  root_shape_id: String,
}

impl EndpointBody {
  fn new(
    body_location: &BodyAnalysisLocation,
    root_shape_id: Option<String>,
    body_commands: impl IntoIterator<Item = SpecCommand>,
    id_generator: &mut impl SpecIdGenerator,
  ) -> Self {
    let body_descriptor = match root_shape_id {
      Some(root_shape_id) => Some(EndpointBodyDescriptor {
        content_type: body_location
          .content_type()
          .expect("root shape id implies a content type to be present")
          .clone(),
        root_shape_id,
      }),
      None => None,
    };

    let mut body = match body_location {
      BodyAnalysisLocation::UnmatchedRequest {
        path_id, method, ..
      } => EndpointBody::Request(EndpointRequestBody {
        body_descriptor,
        path_id: path_id.clone(),
        method: method.clone(),
        commands: body_commands.into_iter().collect(),
      }),
      BodyAnalysisLocation::UnmatchedResponse {
        status_code,
        path_id,
        method,
        ..
      } => EndpointBody::Response(EndpointResponseBody {
        body_descriptor,
        path_id: path_id.clone(),
        method: method.clone(),
        commands: body_commands.into_iter().collect(),
        status_code: *status_code,
      }),
      _ => panic!("EndpointBody should only be created for unmatched responses and requests"),
    };

    body.append_endpoint_commands(id_generator);

    body
  }

  fn append_endpoint_commands(&mut self, ids: &mut impl SpecIdGenerator) {
    match self {
      EndpointBody::Request(request_body) => {
        let request_id = ids.request();
        request_body
          .commands
          .push(SpecCommand::from(EndpointCommand::add_request(
            request_id.clone(),
            request_body.path_id.clone(),
            request_body.method.clone(),
          )));

        if let Some(body_descriptor) = &request_body.body_descriptor {
          request_body
            .commands
            .push(SpecCommand::from(EndpointCommand::set_request_body_shape(
              request_id,
              body_descriptor.root_shape_id.clone(),
              body_descriptor.content_type.clone(),
              false,
            )));
        }
      }
      EndpointBody::Response(response_body) => {
        let response_id = ids.response();
        response_body.commands.push(SpecCommand::from(
          EndpointCommand::add_response_by_path_and_method(
            response_id.clone(),
            response_body.path_id.clone(),
            response_body.method.clone(),
            response_body.status_code.clone(),
          ),
        ));

        if let Some(body_descriptor) = &response_body.body_descriptor {
          response_body
            .commands
            .push(SpecCommand::from(EndpointCommand::set_response_body_shape(
              response_id,
              body_descriptor.root_shape_id.clone(),
              body_descriptor.content_type.clone(),
              false,
            )));
        }
      }
    };
  }
}

#[cfg(test)]
mod test {
  use super::*;
  use crate::interactions::InteractionDiffResult;
  use crate::learn_shape::observe_body_trails;
  use crate::state::body::BodyDescriptor;
  use insta::assert_debug_snapshot;
  use serde_json::json;

  #[test]
  fn undocumented_bodies_can_aggregate_analysis_results_with_array_items() {
    let body = BodyDescriptor::from(json!({
      "items": [132, "string-array-item"],
      "other-field": true
    }));

    let analysis_result = BodyAnalysisResult {
      body_location: BodyAnalysisLocation::MatchedResponse {
        response_id: String::from("test-response-1"),
        content_type: Some(String::from("application/json")),
        status_code: 200,
      },
      trail_observations: observe_body_trails(body),
    };

    let mut projection = LearnedUndocumentedBodiesProjection::default();

    projection.apply(analysis_result.clone());

    // dbg!(&projection);
    let aggregated_result = projection
      .observations_by_body_location
      .get(&analysis_result.body_location)
      .unwrap();

    let array_trail = JsonTrail::empty().with_object_key(String::from("items"));

    let items_values = aggregated_result
      .values_by_trail
      .get(&array_trail.with_array_item(0))
      .expect("should have learned values for items array");
    assert!(items_values.was_number && items_values.was_string);
    assert!(
      aggregated_result
        .values_by_trail
        .get(&array_trail.with_array_item(1))
        .is_none(),
      "undocumented endpoints projection aggregates normalized body results",
    );

    let mut aggregated_trails: Vec<_> = aggregated_result.trails().collect();
    aggregated_trails.sort();

    assert_debug_snapshot!(
      "undocumented_bodies_can_aggregate_analysis_results_with_array_items__aggregated_trails",
      aggregated_trails
    );
  }

  #[test]
  fn undocumented_bodies_generates_commands_for_responses_without_bodies() {
    let analysis_result = BodyAnalysisResult {
      body_location: BodyAnalysisLocation::UnmatchedResponse {
        content_type: None,
        path_id: String::from("path-1"),
        method: String::from("DELETE"),
        status_code: 204,
      },
      trail_observations: observe_body_trails(None),
    };

    let mut test_id_generator = TestIdGenerator::default();
    let mut projection = LearnedUndocumentedBodiesProjection::default();

    projection.apply(analysis_result);

    let endpoint_bodies = projection
      .into_endpoint_bodies(&mut test_id_generator)
      .collect::<Vec<_>>();

    assert_debug_snapshot!(
      "undocumented_bodies_generates_commands_for_responses_without_bodies__endpoint_bodies",
      endpoint_bodies
    );
  }

  #[derive(Debug, Default)]
  struct TestIdGenerator {
    counter: usize,
  }

  impl SpecIdGenerator for TestIdGenerator {
    fn generate_id(&mut self, _prefix: &str) -> String {
      let id = format!("test-id-{}", self.counter);
      self.counter += 1;
      id
    }
  }
}
