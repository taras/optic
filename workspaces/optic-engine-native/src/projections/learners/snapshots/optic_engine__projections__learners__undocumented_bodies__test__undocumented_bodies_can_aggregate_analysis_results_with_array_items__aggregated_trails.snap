---
source: workspaces/optic-engine-native/src/projections/learners/undocumented_bodies.rs
expression: aggregated_trails
---
[
    JsonTrail {
        path: [],
    },
    JsonTrail {
        path: [
            JsonObjectKey {
                key: "items",
            },
        ],
    },
    JsonTrail {
        path: [
            JsonObjectKey {
                key: "items",
            },
            JsonArrayItem {
                index: 0,
            },
        ],
    },
    JsonTrail {
        path: [
            JsonObjectKey {
                key: "other-field",
            },
        ],
    },
]
